src/utils/ensureTimeout.ts

```diff
@@ -0,0 +1,42 @@
+import { CancelToken, isCancelled, RequestConfiguration } from './cancelRequests';
+
+export const ensureTimeout = async (
+  innerFunction: (requestConfig: RequestConfiguration) => Promise<any>,
+  reqConfig?: RequestConfiguration,
+): Promise<any> => {
+  if (!reqConfig) {
+    const innerResult = await innerFunction(reqConfig);
+    return innerResult;
+  }
+
+  const { cancelToken, timeout } = reqConfig;
+
+  if (!timeout) {
+    const innerResult = await innerFunction(reqConfig);
+    return innerResult;
+  }
+
+  if (!cancelToken) {
+    const token = new CancelToken();
+    reqConfig.cancelToken = token;
+  }
+
+  const timer = setTimeout(() => {
+    reqConfig.cancelToken.cancel();
+    clearTimeout(timer);
+  }, timeout);
+
+  try {
+    const resolvedValue = await innerFunction(reqConfig);
+    clearTimeout(timer);
+    return resolvedValue;
+  } catch (e) {
+    if (isCancelled(e)) {
+      clearTimeout(timer);
+      return null;
+    } else {
+      clearTimeout(timer);
+      throw e;
+    }
+  }
```

Comment:
We should still throw an exception if timeout happened, so this should probably be:
```
} catch(e) {
  clearTimeout(timer);
  throw e;
}
```

Reply:
Updated